From 7e5c41c376cd3bd69adcf7b12bfb26bb9659584d Mon Sep 17 00:00:00 2001
From: Craig Chapel <cchapel@codeaurora.org>
Date: Wed, 13 Feb 2013 10:14:08 -0700
Subject: [PATCH] RFC: Add reviewer to label query

This change supports query for reviewer in a label
eg  label:"CodeReview=+1~cchapel"
Currently it only works within quotes and with label:
(default shortcuts such as verified+1 don't work)

Change-Id: I39a1000712743a6ed238e70d45c0ccc1eec50031
---
 Documentation/user-search.txt                      | 14 ++++++++++++++
 .../server/query/change/ChangeQueryBuilder.java    | 22 +++++++++++++++++++---
 .../gerrit/server/query/change/LabelPredicate.java | 11 ++++++++++-
 3 files changed, 43 insertions(+), 4 deletions(-)

diff --git a/Documentation/user-search.txt b/Documentation/user-search.txt
index 3bd437f..65c215f 100644
--- a/Documentation/user-search.txt
+++ b/Documentation/user-search.txt
@@ -333,6 +333,9 @@ A label name is any of the following:
 * The one or two character abbreviation shown in the column header
   of change list pages.  Example: `label:R` or `label:V`.
 
+* The category name followed by a '~' followed by a reviewer where the category name
+  and reviewer combination is enclosed in double quotes.
+
 A label name must be followed by a score, or an operator and a score.
 The easiest way to explain this is by example.
 
@@ -359,10 +362,21 @@ Scores of +2 are not matched, even though they are higher.
 +
 Matches changes with either a +1, +2, or any higher score.
 
+'label:"CodeReview=2~jsmith"'::
++
+Matches changes with a +2 code review where the reviewer is jsmith
+NOTE: Tilde is used to separate the code review value from the reviewer.
+      Currently it only works within quotes and with label:
+      (default shortcuts such as Verified+1 don't work).
+
 `label:CodeReview<=-1`::
 +
 Matches changes with either a -1, -2, or any lower score.
 
+'CodeReview=+2'::
++
+Matches change with a score of +2.
+
 `is:open CodeReview+2 Verified+1 -Verified-1 -CodeReview-2`::
 +
 Matches changes that are ready to be submitted.
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/query/change/ChangeQueryBuilder.java b/gerrit-server/src/main/java/com/google/gerrit/server/query/change/ChangeQueryBuilder.java
index 1d34c24..215272c 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/query/change/ChangeQueryBuilder.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/query/change/ChangeQueryBuilder.java
@@ -300,9 +300,21 @@ public class ChangeQueryBuilder extends QueryBuilder<ChangeData> {
   }
 
   @Operator
-  public Predicate<ChangeData> label(String name) {
+  public Predicate<ChangeData> label(String name) throws QueryParseException,
+      OrmException {
+    Set<Account.Id> accounts = null;
+
+    String splitReviewer[] = name.split("~");
+    name = splitReviewer[0];
+
+    if (splitReviewer.length > 2) {
+      throw new QueryParseException("more than one user specified");
+    } else if (splitReviewer.length == 2) {
+      accounts = parseAccount(splitReviewer[1]);
+    }
+
     return new LabelPredicate(args.changeControlGenericFactory,
-        args.userFactory, args.dbProvider, args.approvalTypes, name);
+        args.userFactory, args.dbProvider, args.approvalTypes, name, accounts);
   }
 
   @Operator
@@ -514,7 +526,11 @@ public class ChangeQueryBuilder extends QueryBuilder<ChangeData> {
       }
 
     } else if (PAT_LABEL.matcher(query).find()) {
-      return label(query);
+      try {
+        return label(query);
+      } catch (OrmException err) {
+        throw error("Cannot lookup user", err);
+      }
 
     } else {
       // Try to match a project name by substring query.
diff --git a/gerrit-server/src/main/java/com/google/gerrit/server/query/change/LabelPredicate.java b/gerrit-server/src/main/java/com/google/gerrit/server/query/change/LabelPredicate.java
index afec452..02a7234 100644
--- a/gerrit-server/src/main/java/com/google/gerrit/server/query/change/LabelPredicate.java
+++ b/gerrit-server/src/main/java/com/google/gerrit/server/query/change/LabelPredicate.java
@@ -26,6 +26,7 @@ import com.google.gerrit.server.IdentifiedUser;
 import com.google.gerrit.server.project.ChangeControl;
 import com.google.gerrit.server.project.NoSuchChangeException;
 import com.google.gerrit.server.query.OperatorPredicate;
+import com.google.gerrit.server.query.QueryParseException;
 import com.google.gwtorm.server.OrmException;
 import com.google.inject.Provider;
 
@@ -34,6 +35,9 @@ import java.util.Set;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
+import java.util.Collections;
+import java.util.Set;
+
 class LabelPredicate extends OperatorPredicate<ChangeData> {
   private static enum Test {
     EQ {
@@ -117,14 +121,16 @@ class LabelPredicate extends OperatorPredicate<ChangeData> {
   private final ApprovalCategory category;
   private final String permissionName;
   private final int expVal;
+  private final Set<Account.Id> accounts;
 
   LabelPredicate(ChangeControl.GenericFactory ccFactory,
       IdentifiedUser.GenericFactory userFactory, Provider<ReviewDb> dbProvider,
-      ApprovalTypes types, String value) {
+      ApprovalTypes types, String value, Set<Account.Id> accounts) {
     super(ChangeQueryBuilder.FIELD_LABEL, value);
     this.ccFactory = ccFactory;
     this.userFactory = userFactory;
     this.dbProvider = dbProvider;
+    this.accounts = accounts;
 
     Matcher m1 = Pattern.compile("(=|>=|<=)([+-]?\\d+)$").matcher(value);
     Matcher m2 = Pattern.compile("([+-]\\d+)$").matcher(value);
@@ -152,6 +158,9 @@ class LabelPredicate extends OperatorPredicate<ChangeData> {
     final Set<Account.Id> allApprovers = new HashSet<Account.Id>();
     final Set<Account.Id> approversThatVotedInCategory = new HashSet<Account.Id>();
     for (PatchSetApproval p : object.currentApprovals(dbProvider)) {
+      if (accounts != null && ! accounts.contains(p.getAccountId())) {
+        continue;
+      }
       allApprovers.add(p.getAccountId());
       if (p.getCategoryId().equals(category.getId())) {
         approversThatVotedInCategory.add(p.getAccountId());
-- 
1.8.1.2

